ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,273 @@
+
+
+### Codecs
+
+And a payload codec is expected to implement this interface:
+
+```ruby
+interface _PayloadCodec
+  # Takes an array of Payload objects and returns an array of encoded Payload object
+  def encode: (Array[Temporal::Payload]) -> Array[Temporal::Payload]
+
+  # Takes an array of Payloads object and returns an array of decoded Payload object
+  def decode: (Array[Temporal::Payload]) -> Array[Temporal::Payload]
+end
+```
+
+
+### Configuring client/worker
+
+In order to use the defined converters and codecs, these need to be supplied to a client and worker.
+
+```ruby
+converter = Temporal::Converters::Composite.new(
+  Temporal::Converters::Null,
+  Temporal::Converters::Bytes,
+  MyProtobufConverter.new
+)
+codecs = [EncryptionCodec.new(KEY_ID), CompressionCodec]
+
+connection = Temporal::Connection.new('http://localhost:7233')
+client = Temporal::Client.new(
+  connection,
+  'my-namespace',
+  payload_converter: converter,
+  payload_codecs: codecs
+)
+
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [...],
+  activities: [...],
+  payload_converter: converter,
+  payload_codecs: codecs
+)
+```
+
+*This does indeed involve some duplication and we might simplify this initialization later
+on. But it's outside the scope for this proposal.*
+
+Notes:
+
+- Converters and codecs both can be classes and class instances as long as the given object
+  implements the required interface
+- This API is still being figured out to accommodate future features and configuration params
+
+
+## Data conversion
+
+All the previously outlined modules are required to facilitate the process of data conversion.
+Here's the diagram representing it:
+
+![Data conversion](./images/data-conversion.png)
+
+This describes the process that every piece of data (workflow/activity args and return values,
+signal inputs, query results, headers, memos, etc) goes through getting transformed from the
+user-land object to something that can be transmitted over the wire.
+
+*NOTE: As mentioned previously, the "Type hints" are optional and will be omitted altogether unless
+defined.*
+
+The reverse process is applied when converting the data from a protobuf received over the wire back
+to a use-land object.
+
+
+## Errors
+
+Here's the Error hierarchy that is planned for the Ruby SDK:
+
+```ruby
+module Temporal
+  # top level error superclass
+  class Error < StandardError
+
+  # superclass for all errors within the SDK itself
+  class InternalError < Error
+
+  class ClientError < InternalError # errors specific to Client
+  class MissingNamespace < ClientError
+  ...
+
+  class WorkerError < InternalError # errors specific to Worker
+  class TooManyStickyWorkflows < WorkerError
+  ...
+
+  # superclass for connection and network errors
+  class NetworkError < Error
+
+  class TimeoutError < NetworkError
+  class HostUnreachable < NetworkError
```

Comment:
I am not sure you can reliably capture these. I think it'd be more reasonable, instead of trying to guess at all the ones you can catch, to instead just have an `RPCError` from Core that contains the gRPC status code and message. Then we can specialize as needed. I think you are overthinking this and adding way too many errors.

Reply:
Good point, let me simplify this